llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp

```diff
@@ -0,0 +1,251 @@
+  }
+  SmallPtrSet<const Use *, 4> Uses;
+  collectUsesSkipThroughCasts(LocalVar, Uses);
+
+  for (const Use *U : Uses) {
+    Value *V = U->getUser();
+
+    if (auto *StI = dyn_cast<StoreInst>(V)) {
+      constexpr int StoreInstValueOperandIndex = 0;
+
+      if (U != &StI->getOperandUse(StoreInst::getPointerOperandIndex())) {
+        assert(U == &StI->getOperandUse(StoreInstValueOperandIndex));
+        // this is double indirection - not supported
+        llvm_unreachable("unsupported data flow pattern for invoke_simd 11");
+      }
+      V = stripCasts(StI->getValueOperand());
+
+      if (auto *LI = dyn_cast<LoadInst>(V)) {
+        // A value loaded from another address is stored at this address -
+        // recurse into the other addresss
+        getPossibleStoredVals(LI->getPointerOperand(), Vals);
+      } else {
+        Vals.insert(V);
+      }
+      continue;
+    }
+    if (const auto *CI = dyn_cast<CallInst>(V)) {
+      // only __builtin_invoke_simd is allowed, otherwise the pointer escapes
+      if (!getInvokeeIfInvokeSimdCall(CI)) {
+        llvm_unreachable("unsupported data flow pattern for invoke_simd 12");
+      }
+      continue;
+    }
+    if (const auto *LI = dyn_cast<LoadInst>(V)) {
+      // LoadInst from this addr is OK, as it does not affect what can be stored
+      // through the addr
+      continue;
+    }
+    llvm_unreachable("unsupported data flow pattern for invoke_simd 13");
+  }
+}
+
+// Example1 (function is direct argument to _Z21__builtin_invoke_simd):
+// %call6.i = call spir_func float @_Z21__builtin_invoke_simd...(
+//   <16 x float> (float addrspace(4)*, <16 x float>, i32)* %28, <== function
+//   pointer float addrspace(4)* %arg1, float %arg2, i32 %arg3)
+//
+// Example 2 (invoke_simd's target function pointer flows through IR):
+// %"fptr_t = <16 x float> (float addrspace(4)*, <16 x float>, i32)*
+// ...
+// %fa_as0 = alloca %fptr_t
+// ...
+// %fa = addrspacecast %fptr_t* %fa_as0 to %fptr_t addrspace(4)*
+// ...
+// store %fptr_t @__SIMD_CALLEE, %fptr_t addrspace(4)* %fa
+// ...
+// %f = load %fptr_t, %fptr_t addrspace(4)* %fa
+// ...
+// %res = call spir_func float @_Z21__builtin_invoke_simd...(
+//   %fptr_t %f, <== function pointer
+//  float addrspace(4)* %arg1,
+//  float %arg2,
+//  i32 %arg3)
+//
+bool processInvokeSimdCall(CallInst *CI) {
+  Value *V = getInvokeeIfInvokeSimdCall(CI);
+
+  if (!V) {
+    llvm_unreachable(("bad use of " + Twine(INVOKE_SIMD_PREF)).str().c_str());
+  }
+  Function *SimdF = nullptr;
+  bool Modified = false;
+
+  if (!(SimdF = dyn_cast<Function>(V))) {
+    LoadInst *LI = dyn_cast<LoadInst>(stripCasts(V));
+
+    if (!LI) {
+      llvm_unreachable("unsupported data flow pattern for invoke_simd 0");
+    }
+    ValueSet Vals;
+    getPossibleStoredVals(LI->getPointerOperand(), Vals);
+
+    if (Vals.size() != 1 || !(SimdF = dyn_cast<Function>(*Vals.begin()))) {
+      llvm_unreachable("unsupported data flow pattern for invoke_simd 1");
+    }
```

Comment:
```suggestion
    if (Vals.size() != 1 || !isa<Function>(*Vals.begin())) {
      llvm_unreachable("unsupported data flow pattern for invoke_simd 1");
    }
    SimdF = cast<Function>(*Vals.begin());
```

To me, that seems more readable (even though it takes more lines)